### PR TITLE
fix: default pair for unsupported assets

### PR DESCRIPTION
**Pull request by elmutt**
## Description

Buy asset on the default pair defaults to whatever page the user is on. Some assets (foxy) are unsupported by the swapper and cause problems when visiting those pages.

We now check to make sure we can support the pair and default to fox otherwise

Also check to make sure swapper exists before making swapper calls

## Notice

- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Risk

Low risk, fixes foxy asset page

## Testing

Go to foxy page and see that trade area isnt forever loading a quote.

## Screenshots (if applicable)
Screen Shot 2022-05-20 at 10 41 25 AM

### CODE REVIEW

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 1]
```diff
@@ -35,13 +35,18 @@ export const useTradeRoutes = (
...
+      // check pair is valid
+      const isSupportedPair =
+        swapperManager.getSwappersByPair({
+          buyAssetId,
+          sellAssetId,
+        }).length > 0
+
+      const supportedBuyAssetId = isSupportedPair ? buyAssetId : defaultBuyAssetId
+
+      const buyAsset = assets[supportedBuyAssetId as string]
```
Comment (0xdef1cafe) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:49]: plz fix

**Review by 0xdef1cafe — CHANGES_REQUESTED** [review]

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts (hunk 1 above)
Comment (elmutt) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts]: done

**Review by theoboldfrazier — APPROVED** [review]
👍

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 2]
```diff
@@ -32,16 +32,35 @@ export const useTradeRoutes = (
...
+      // TODO update swapper to have an official way to validate a pair is supported.
+      // This works for now
+      const isSupportedPair = await (async () => {
+        try {
+          if (bestSwapper) {
+            await bestSwapper.getUsdRate({ ...assets[buyAssetToCheckId] })
+            return true
+          }
+        } catch (e) {}
+        return false
```
Comment (gomesalexandre) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:58]: return in catch block?

**Review by 0xdef1cafe — APPROVED** [review]

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Works on the first render of the trade component, but if going from one asset to another without the trade component unmounting (by e.g searching and clicking on another asset), the previous pair is still shown. I haven't been able to consistently reproduce it, but from what I noticed this happens:
- When going from an asset to another asset that was previously loaded
- When going to a valid pair to an unsupported asset (e.g FOXy), which will keep the previous pair loaded instead of falling back to the default

The issue is that we're only having `feeAsset` as a dependency, but not the route assetId, this fixes it:

```diff
diff --git a/src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts b/src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts
index 68644a24..63b02405 100644
--- a/src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts
+++ b/src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts
@@ -84 +84 @@ export const useTradeRoutes = (
-  }, [assets, feeAsset]) // eslint-disable-line react-hooks/exhaustive-deps
+  }, [assets, feeAsset, routeBuyAssetId]) // eslint-disable-line react-hooks/exhaustive-deps
```

**Review by gomesalexandre — APPROVED** [review]
Tested again, LGTM now! Swapper is always re-rendering with the current assetId or default 🍬

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 3]
```diff
@@ -14,14 +14,14 @@ import { useSwapper } from '../useSwapper/useSwapper'
 const ETHEREUM_ASSET_ID = 'eip155:1/slip44:60'
 
 export const useTradeRoutes = (
-  defaultBuyAssetId?: AssetId,
+  routeBuyAssetId?: AssetId,
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:17]: This name trips me out. `buyAssetIdRoute`?

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 4]
```diff
@@ -32,16 +32,35 @@ export const useTradeRoutes = (
...
+      const sellAsset = assets[defaultSellAssetId]
 
-      // ugly hack until we add proper error handling in another PR soon
-      const buyAsset =
-        assets[
-          defaultBuyAssetId?.startsWith('eip155:1') && defaultBuyAssetId !== 'eip155:1/slip44:60'
-            ? defaultBuyAssetId
-            : buyAssetId
-        ]
+      const preBuyAssetToCheckId = routeBuyAssetId ?? defaultBuyAssetId
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:38]: This name confuses me.

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts [hunk 5]
```diff
@@ -32,16 +32,35 @@ export const useTradeRoutes = (
...
+
+      // make sure the same buy and sell assets arent selected
+      const buyAssetToCheckId =
```
Comment (0xApotheosis) [inline src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:41]: As does this name.

**Review by 0xApotheosis — APPROVED** [review]
All works as expected, couldn't see any regressions.

A few non-blocking comments on var names that I find confusing.

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts (hunk 3 above)
Comment (elmutt) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:17]: the buy asset id is from the route ultimately. defi suggested this name

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts (hunk 5 above)
Comment (elmutt) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:41]: we use this asset id to ultimely do the check if its supported by the swapper using usdRate

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts (hunk 4 above)
Comment (elmutt) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:38]: Had to define this variable to do the logic on line 41 without using the let pattern

File: src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts (hunk 3 above)
Comment (0xApotheosis) [reply src/components/Trade/hooks/useTradeRoutes/useTradeRoutes.ts:17]: Ah, that's the origin - got it.

**Pull request merged**